x-pack/filebeat/input/etw/input.go

```diff
@@ -0,0 +1,281 @@
+	var once sync.Once
+
+	// Create an error channel to communicate errors from the goroutine
+	errChan := make(chan error, 1)
+
+	defer func() {
+		once.Do(e.Close)
+		e.log.Info(inputName + " input stopped")
+	}()
+
+	// eventReceivedCallback processes each ETW event
+	eventReceivedCallback := func(record *etw.EventRecord) uintptr {
+		if record == nil {
+			e.log.Error("received null event record")
+			return 1
+		}
+
+		e.log.Debugf("received event %d with length %d", record.EventHeader.EventDescriptor.Id, record.UserDataLength)
+
+		data, err := etw.GetEventProperties(record)
+		if err != nil {
+			e.log.Errorf("failed to read event properties: %w", err)
+			return 1
+		}
+
+		evt := buildEvent(data, record.EventHeader, e.etwSession, e.config)
+		publisher.Publish(evt)
+
+		return 0
+	}
+
+	// Set the callback function for the ETW session
+	e.etwSession.Callback = eventReceivedCallback
+
+	// Start a goroutine to consume ETW events
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		e.log.Debug("starting to listen ETW events")
+		if err = e.operator.startConsumer(e.etwSession); err != nil {
+			errChan <- fmt.Errorf("failed to start consumer: %w", err) // Send error to channel
+			return
+		}
+		e.log.Debug("stopped to read ETW events from session")
+		errChan <- nil
+	}()
+
+	// We ensure resources are closed when receiving a cancelation signal
+	go func() {
+		<-ctx.Cancelation.Done()
+		once.Do(e.Close)
+	}()
+
+	wg.Wait() // Ensure all goroutines have finished before closing
+	close(errChan)
+	if err, ok := <-errChan; ok && err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// buildEvent builds the winlog object.
+func buildEvent(data map[string]any, h etw.EventHeader, session *etw.Session, cfg config) beat.Event {
+
+	winlog := map[string]any{
+		"activity_guid": h.ActivityId.String(),
+		"channel":       fmt.Sprintf("%d", h.EventDescriptor.Channel),
+		"event_data":    data,
+		"flags":         fmt.Sprintf("%d", h.Flags),
+		"keywords":      fmt.Sprintf("%d", h.EventDescriptor.Keyword),
+		"opcode":        fmt.Sprintf("%d", h.EventDescriptor.Opcode),
+		"process_id":    fmt.Sprintf("%d", h.ProcessId),
+		"provider_guid": h.ProviderId.String(),
+		"session":       session.Name,
+		"task":          fmt.Sprintf("%d", h.EventDescriptor.Task),
+		"thread_id":     fmt.Sprintf("%d", h.ThreadId),
+		"version":       fmt.Sprintf("%d", h.EventDescriptor.Version),
+	}
+
+	// Include provider GUID if not available in event header
+	zeroGUID := "{00000000-0000-0000-0000-000000000000}"
+	if winlog["provider_guid"] == zeroGUID {
+		winlog["provider_guid"] = session.GUID.String()
+	}
+
+	// Define fields map with Windows data and ECS mapping
+	fields := mapstr.M{
+		"winlog":         winlog,
+		"event.code":     fmt.Sprintf("%d", h.EventDescriptor.Id),
+		"event.created":  time.Now(),
+		"event.kind":     "event",
+		"event.severity": h.EventDescriptor.Level,
+	}
+
+	// Mapping from Level to Severity
+	levelToSeverity := map[uint8]string{
```

Comment:
This mapping should be declared as a global. There's no need to create it every invocation.